### ISSUE TITLE
SA1633 in generated code in ASP.NET Core Web Application

### ISSUE THREAD

**Issue by CWalford-Quest** [DotNetAnalyzers/StyleCopAnalyzers#2493]
I have an issue where using the StyleCop Analyzers with ASP.NET Core reports errors in Release builds for generated code under the _obj_ folder. This problem is seen both in Visual Studio (2017) and building on the command line (using dotnet version 1.0.4).

The generated files are AssemblyInfo.cs and UserSecretsAssemblyInfo.cs. The generation of AssemblyInfo.cs can be prevented by adding _<GenerateAssemblyInfo>false</GenerateAssemblyInfo>_ to the project file. I don't know whether the generation of UserSecretsAssemblyInfo.cs can be prevented and it probably shouldn't be if user secrets are being used.

Steps to recreate below (project to replicate issue attached):

- Create new ASP.NET Core Web Application (.NET Core)
- Add the following lines to the project file

```
	<PropertyGroup>
		<UserSecretsId>UserSecretsKey</UserSecretsId>
		<TreatWarningsAsErrors>true</TreatWarningsAsErrors>
	</PropertyGroup>
	
	<ItemGroup>
		<PackageReference Include="Microsoft.Extensions.Configuration.UserSecrets" Version="1.1.2" />
		<PackageReference Include="StyleCop.Analyzers" Version="1.1.0-beta004" PrivateAssets="All" />
	</ItemGroup>
	
	<PropertyGroup>
		<CodeAnalysisRuleSet>StyleCop.ruleset</CodeAnalysisRuleSet>
	</PropertyGroup>
	<ItemGroup>
		<AdditionalFiles Include="StyleCop.json" />
	</ItemGroup>
```

- Create file StyleCop.json in the project folder:

```
{
  "$schema": "https://raw.githubusercontent.com/DotNetAnalyzers/StyleCopAnalyzers/master/StyleCop.Analyzers/StyleCop.Analyzers/Settings/stylecop.schema.json",
  "settings": {
    "documentationRules": {
      "companyName": "My Company",
      "copyrightText": "Some lines\nto go in the\nheader"
    }
  }
}
```

- Create file StyleCop.ruleset in the project folder:

```
<?xml version="1.0" encoding="utf-8"?>
<RuleSet Name="StyleCopeRules" Description="StyleCopeRules custom ruleset" ToolsVersion="15.0">
  <IncludeAll Action="Warning" />
  <Rules AnalyzerId="StyleCop.Analyzers" RuleNamespace="StyleCop.Analyzers">
	<!-- SA0001: XmlCommentAnalysisDisabled -->
    <Rule Id="SA0001" Action="None" />
  </Rules>
  <Rules AnalyzerId="StyleCop.Analyzers" RuleNamespace="StyleCop.Analyzers">
	<!-- SA1600: ElementsMustBeDocumented -->
    <Rule Id="SA1600" Action="None" />
  </Rules>
</RuleSet>
```

- Modify Startup.cs Startup method to read:

```
        public Startup(IHostingEnvironment env)
        {
            var builder = new ConfigurationBuilder()
                .SetBasePath(env.ContentRootPath)
                .AddJsonFile("appsettings.json", optional: false, reloadOnChange: true)
                .AddJsonFile($"appsettings.{env.EnvironmentName}.json", optional: true);

            if (env.IsDevelopment())
            {
                builder.AddUserSecrets("UserSecretsKey");

                // This will push telemetry data through Application Insights pipeline faster, allowing you to view results immediately.
                builder.AddApplicationInsightsSettings(developerMode: true);
            }

            builder.AddEnvironmentVariables();

            Configuration = builder.Build();
        }
```
		
- Fix StyleCop issues in VisualStudio

- Switch configuration to Release and build project or run "dotnet build --configuration Release" from the command line. I see the following errors:

```
obj\Release\netcoreapp1.0\UserSecretsAssemblyInfo.cs(1,1): error SA1633: The file header XML is invalid. [C:\Code\WebApplication2\WebApplication2.csproj]
obj\Release\netcoreapp1.0\WebApplication2.AssemblyInfo.cs(1,1): error SA1633: The file header XML is invalid. [C:\Code\WebApplication2\WebApplication2.csproj]
```

[WebApplication2.zip](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/files/1152084/WebApplication2.zip)

**Comment by sharwell** [comment]
@CWalford-Quest

I'm curious what led to this:

```xml
<Rule Id="SA0001" Action="None" />
```

We renumbered this diagnostic and gave it a stronger error message since users in the past were disabling it. The impact of this configuration is so pronounced that I'm starting to think we need to remove the ability to change the severity of this rule.

**Comment by sharwell** [comment]
Also, this is rather unpleasant for daily work:

```xml
<TreatWarningsAsErrors>true</TreatWarningsAsErrors>
```

I would recommend gating this so users working in Visual Studio will see warnings instead of errors, allowing for a much more efficient iterative <kbd>F5</kbd> workflow. [We use](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/abd2b1c71d0a23b41ab6f140eda907e153894085/StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/StyleCop.Analyzers.CodeFixes.csproj#L36-L39) the following configuration for this project:

```xml
<PropertyGroup Condition="'$(BuildingInsideVisualStudio)' != 'true'">
  <!-- Ideally this is always enabled, but that tends to hurt developer productivity -->
  <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
</PropertyGroup>
```

**Comment by CWalford-Quest** [comment]
I didn't want to have to go through and add documentation to all the methods for the purpose of recreating the issue. I thought it was easier just to disable that rule.

**Comment by sharwell** [comment]
I had a feeling that was the case. :smile: Here's a property group you can use to achieve this without the downsides:

```xml
<PropertyGroup>
  <GenerateDocumentationFile>True</GenerateDocumentationFile>
  <NoWarn>$(NoWarn);CS1591</NoWarn>
</PropertyGroup>
```

The SA0001 mostly stood out because it was a *really* commonly reported problem here. I see up to 30 bug reports where the root cause turned out to be someone disabling this rule.

**Comment by sharwell** [comment]
@CWalford-Quest This issue was caused by a bug in the dotnet SDK (or MSBuild) which has since been fixed. If you update to a more recent version of the tools, the generated files will contain a proper `<auto-generated>` element in the file header which will exclude them from analysis.

**Comment by sharwell** [comment]
Duplicate of #2314

**Comment by CWalford-Quest** [comment]
I'm already on version 1.0.4 of the dotnet SDK. Are you suggesting that I need to install a 2.0 SDK, if not could you be specific about what tools I need to update. Thanks.

**Comment by sharwell** [comment]
@CWalford-Quest Weird. I can reproduce this issues with `dotnet build` (2.0.0-preview2-006461), but not with Visual Studio 2017 15.3 Preview 4.

**Comment by CWalford-Quest** [comment]
The contents of the generated files seems to be different on the command line and in Visual Studio, see attached files.

[BuiltInVisualStudio-UserSecretsAssemblyInfo.cs.txt](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/files/1152724/BuiltInVisualStudio-UserSecretsAssemblyInfo.cs.txt)
[BuiltOnCommandLine-UserSecretsAssemblyInfo.cs.txt](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/files/1152725/BuiltOnCommandLine-UserSecretsAssemblyInfo.cs.txt)

**Comment by sharwell** [comment]
@CWalford-Quest Yes, that's what I'm seeing. I believe a future release of the dotnet tool will produce code similar to what you're seeing from Visual Studio currently.

**Comment by CWalford-Quest** [comment]
So do I take it that there is currently no resolution for this issue other than wait for a new version of dotnet?

**Comment by sharwell** [comment]
You may be able to disable assembly info generation during the build, and instead adding the file manually:

```xml
<GenerateAssemblyInfo>False</GenerateAssemblyInfo>
```

**Comment by CWalford-Quest** [comment]
Unfortunately disabling the `GenerateAssemblyInfo` setting only prevents the AssemblyInfo.cs file from being generated, the UserSecretsAssemblyInfo.cs is still generated. Thanks for your quick response and help with this issue.
